Passionfruit

22a opened the pull request:
Opening to see how many cucumbers I have snapped.  🥒

This PR:
* **Refactors request chain out into its own class**
* **Displays more of the information about DDI provided by the API**
* **Adds Tab UI**

Tab UI looks like this:
<img width="734" alt="screen shot 2017-04-01 at 04 31 24" src="https://cloud.githubusercontent.com/assets/7144173/24575047/2e1575e0-1694-11e7-90fa-e955cd8e15e6.png">
and implements a few possibly useful things like the error tab pulling focus when there's an error and reverting back to the analysis tab otherwise, and automatic numbering on the warning tab (which is also possible to implement for the error tab but atm I've hard coded that to `1` because one error breaks our request chain thus preventing a second error).

It also moves a lot of the `if ( things_we_want_to_display.length > 0) {` logic into the view module.

22a committed:
request chain refactor

22a committed:
made drug display less stringy

22a committed:
modified ddi display, adds category, harm, and spacing info

22a committed:
implements tab UI

22a commented:
Nice, many sheared vegetables, I'll worry about those tests tomorrow, from what I can tell they're mostly about the warning panels not being visible when there's no errors (as they're a tab click away now).

Also I forgot to include a screenshot of the new DDI display:
<img width="742" alt="screen shot 2017-04-01 at 04 39 01" src="https://cloud.githubusercontent.com/assets/7144173/24575115/67371800-1695-11e7-82ab-dacad864c3d2.png">
 the wrapper around the ddi info changes from green to red depending on the `harmful` attribute.

As always, if you don't like this feel free to change it, nothing's set in stone.

Also I didn't know if the `enclosing_constructs` attribute is relevant to some feature so I didn't display it.

c-brenn commented:
:wow: Looks very nice.

Do you think we should display the line numbers of the drugs in a DDI? They're in the `/api/ddis` response iirc.

c-brenn commented:
Colour coding DDIs is a really good idea.

One strange nuance of his DDI categorisations is `alternative` DDIS. An `alternative` DDI is when two arms of a branch contain drugs that interact when taken together. Because they're in a branch they can't be taken together and so should never actually interact. The customer wants us to highlight them anyway.

I'd suggest maybe orange for `alternative` DDIS if they are harmful. And green otherwise.

All other DDIs are harmful if the participating drugs are harmful.

c-brenn commented on panacea/web/static/js/app.js:
```diff
@@ -1,104 +1,38 @@
-    await handleDrugsResponse(await Request.drugs(new FormData(this)));
+    retrievePML(pml => {
+      const requestChain = new RequestChain(pml);
+      requestChain.start(formData);
```
Do we need to `await` the request Chain?

Perhaps not, but I'm just remembering when we introduced it without awaits and it messed up the spinner.

22a replied:
We **do** need await, I didn't notice but our loading icon wasn't showing up at all because of synchronicity issues. Great spot, I have just pushed a fix 🥇

c-brenn commented on panacea/web/static/js/view.js:
```diff
@@ -58,32 +107,42 @@ const displayWarnings = (container, preambleText, warnings, pml) => {
+    displayWarnings(container, preamble, warnings, pmlLines);
+    showElement(unnamedPanel);
+  } else {
+    // no unnamed properties to display
```
I don't think we'll ever use the `else` branch.

22a replied:
We don't, I put it in there with the idea that we might display some info along the lines that `the thing you're trying to display is empty`. In the spirit of YAGNI I will remove it 👍

c-brenn commented on panacea/web/static/js/view.js:
```diff
@@ -58,32 +107,42 @@ const displayWarnings = (container, preambleText, warnings, pml) => {
+    displayWarnings(container, preamble, warnings, pmlLines);
+    showElement(clashesPanel);
+  } else {
+    // no clashes to display
```
same here.

c-brenn commented on panacea/web/static/js/view.js:
```diff
@@ -100,41 +159,100 @@ const hideDownloadButton = () => {
+  const drugInfoElement = generateDdiInfoSnippet('Drugs Involved', `${ddi.drug_a.label} and ${ddi.drug_b.label}`);
+  const harmfulElement = generateDdiInfoSnippet('Disposition', ddi.harmful ? 'Harmful' : 'Not Harmful');
+  const categoryElement = generateDdiInfoSnippet('Category', ddi.category);
+  const spacingElement = generateDdiInfoSnippet('Spacing', `${ddi.spacing} hours`);
```
I think the spacing is in days rather than hours.

c-brenn reviewed (changes requested):
Nice job. A few questions here and there.

22a committed:
await request chain, fixes loading icon [skip ci]

c-brenn commented on panacea/web/static/js/app.js:
```diff
@@ -1,104 +1,39 @@
+    retrievePML(async (pml) => {
+      const requestChain = new RequestChain(pml);
+      await requestChain.start(formData);
+      View.restoreFileForm();
```
I notice we're calling `View.restoreFileForm()` in the try and catch. Did you consider putting it in a `finally` block?

22a replied:
Good shout, will investigate!

22a replied:
Just tried it, I don't think it will work. The finally block gets executed without waiting for the await in the `retrievePML`, it's a nuance of js' filthy callbacks and how our `retreivePML` function only calls back in a `fileReader`'s `onload` event handler.

c-brenn replied:
Oh I see. Is it possible to move the try catch into the callback?

22a replied:
Great idea, that works perfectly! 💯 / 💯

22a committed:
minor ddi UI changes

22a committed:
relocate try-catch block [skip ci]

this allows us to utilise a finally block without breaking the loading-icon
logic

houli commented on panacea/web/static/js/request-chain.js:
```diff
@@ -0,0 +1,66 @@
+import * as View from './view';
+import * as Request from './request';
+
+export class RequestChain {
```
if you `export default` here you can `import RequestChain from...` without the `{}`

22a replied:
nice one! thanks!

houli commented on panacea/web/static/css/app.css:
```diff
@@ -70,10 +70,82 @@
+  padding: 10px;
+}
+.ddi-info-label {
+    display: table-cell;
```
Somehow this CSS ran away

22a replied:
It did indeed, very peculiar, where's our csslint?

houli commented:
Want to also sneak in `Save PML TX file` instead of download

c-brenn commented:
Do you think we can get rid of the TX? Just `Save PML File`?

houli commented:
It's a _brand_

c-brenn commented:
I'm wondering about the positioning of the download button too. Not a fan of it's current spot at the bottom. Any ideas?

22a committed:
moved save pml button further up [skip ci]

22a committed:
fixed cucumber tests

houli commented:
Oh actually, a test for the numbers yes

c-brenn commented:
Certainly worthy of a mention in the change log too.

houli commented:
The testing documentation needs a sizy overhaul but I will assign myself that

22a committed:
added error and warning badge number tests

22a committed:
add ddi sequencing UI info to changelog

c-brenn commented on doc/CHANGELOG.md:
```diff
@@ -1,6 +1,13 @@
+
+## [Unreleased]
+
+### Added
```
Perhaps `Changed` would be more suitable?

22a replied:
Makes sense, moved.

22a committed:
minor UI modifications, updated changelog

corrected ddi category UI
added error highlighting to syntax error UI
updated the changelog to reflect the syntax error changes

c-brenn commented on doc/CHANGELOG.md:
```diff
@@ -6,7 +6,12 @@ All notable changes to this project will be documented in this file.
+  alternative, and alternative non-DDI
+
+### Changed
+- The syntax error UI now includes contextual line highlighting
```
some mention of the new tabs perhaps?

22a replied:
good shout, forgot about that ...

22a committed:
Merge branch 'iteration-7' into passionfruit

houli committed:
Fix broken tabs

c-brenn commented on panacea/web/static/js/view.js:
```diff
@@ -100,41 +155,122 @@ const hideDownloadButton = () => {
+  errorsBadge.innerHTML = '1';
+  errorsTab.click();
   errorTitleElement.innerHTML = error.title;
   errorTextElement.innerHTML = error.detail;
```
What do you think about sticking `error.detail` in a `<strong>`?

Some other errors (no drugs) do this iirc.

22a replied:
Sounds good to me, added.

22a committed:
rejigged changelog entries, made errors bold

houli commented:
Some of the finest grapefruit on the market, I'm happy

c-brenn commented on panacea/web/static/js/view.js:
```diff
@@ -263,7 +263,7 @@ export const displayError = (error, pml) => {
   errorsTab.click();
   errorTitleElement.innerHTML = error.title;
-  errorTextElement.innerHTML = error.detail;
+  errorTextElement.innerHTML = `<strong>${error.detail}</strong>`;
```
Some of the call sites already wrap the `detail` in a strong or h5. Should probably remove those.

c-brenn reviewed (changes requested):
not so hasty

c-brenn committed:
removes "line X --" from parser messages

22a committed:
fix cucumis sativus tests, rejig error formatting
